device/core/src/pnp/properties.ts

```diff
@@ -0,0 +1,81 @@
+import { JSONValue, JSONObject } from '.';
+import { Twin } from '../twin';
+
+export function generateWritablePropertyResponse(value: JSONValue, ackCode: number, ackDescription: string, ackVersion: number): {
+    value: JSONValue,
+    ac: number,
+    ad: string,
+    av: number
+} {
+    return {
+        value,
+        ac: ackCode,
+        ad: ackDescription,
+        av: ackVersion
+    };
+}
+
+export class ClientPropertyCollection {
+    /**
+     * The object representation of the properties.
+     */
+    backingObject: JSONObject = {};
+
+    constructor(backingObject?: JSONObject) {
+        this.backingObject = backingObject;
+    }
+
+    /**
+     * The $version field of the properties.
+     * Returns NaN if the $version field is undefined or cannot be converted to a number.
+     */
+    get version(): number {
+        return Number(this.backingObject.$version);
+    }
+
+    /**
+     * Assigns the provided value to a specified property for a component.
+     * If no component is specified, the default component is assumed.
+     *
+     * @param {string} componentName The name of the component.
+     * @param {string} propertyName  The name of the property.
+     * @param {JSONValue} value      The value to assign to the property.
+     */
+    setProperty(propertyName: string, value: JSONValue): void;
+    setProperty(componentName: string, propertyName: string, value: JSONValue): void;
+    setProperty(propertyNameOrComponentName: string, valueOrPropertyName: JSONValue, value?: JSONValue): void {
```

Comment:
>propertyNameOrComponentName: [](http://example.com/codeflow?start=16&length=28)

I'm not sure yet and I'm asking the cross-language cabal., but we may want to skip the "component name is optional" overloading and split this method (and `getProperty`) into `setRootProperty` and `setComponentProperty`

Reply:
Resolved over Teams. Keeping it the way it is.